### DISCUSSION TITLE
Is is possible to draw polygon with interior?

### DISCUSSION THREAD

**Discussion by sharon-br** [pymupdf/PyMuPDF#2314] [Q&A] [answered]
Hey,
I'm trying to draw polygons that contain holes in them over an existing provided PDF file.
My idea was to open the provided pdf file, create a new fitz shape for every polygon I have and use the `Shape.draw_polylines` to draw it.
However, each polygon is constructed from an exterior - which is a collection of points that I can pass to the `draw_polylines` function, but it also contains some interiors which are like "holes" in that exterior polygon.

Is there any option to draw such polygons?

I once saw a different approach of generating an SVG string with interior and creating a pdf from that with something like whtmltopdf, but I want a direct approach.

**Comment by JorjMcKie** [comment, marked as answer]
Yes, this is possible using one of the non-zero-winding or even-odd rules. Example:
```python
import fitz
doc=fitz.open()
page=doc.new_page()
r1 = fitz.Rect(100,100,200,200)
r2 = r1 + (20,20,-20,-20)
shape = page.new_shape()
shape.draw_rect(r1)
shape.draw_rect(r2)
shape.finish(fill=(1,1,0),even_odd=True)
shape.commit()
doc.save("x.pdf")
```
Gives you this:
![grafik](https://user-images.githubusercontent.com/8290722/228835171-5876b4c5-911f-4584-9d91-a022840b5245.png)
Important:
The graphic elements you want to include in such a reglement must belong to the scope of the same `Shape.finish()` method - like shown above.

If letting `even_odd` default to `False`, non-zero-winding is used as the rule. In that case, the above effect will not occur if drawing happens in this way, because both figures are drawn in the same direction (i.e. counter-clockwise). To achieve the same thing, you e.g. would have to draw the outer figure anti-clockwise, and the inner one clockwise.

**Comment by JorjMcKie** [comment]
For a better understanding or details, please read the documentation and also maybe consult information sources like Wikipedia.
Ah, and I hope you have sufficient insight now to do the same for polygons?

**Comment by JorjMcKie** [comment]
Another example:
Draw a polyline in clockwise manner, then draw a circle inside in default (anti-clockwise) manner.
Use default rule "non-zero-winding".
```python
r1 = fitz.Rect(100,100,200,200)
shape=page.new_shape()
shape.draw_polyline((r1.tl, r1.tr, r1.br, r1.bl, r1.tl))
shape.draw_circle((r1.tl+r1.br)/2, 50)
shape.finish(fill=(1,1,0))
shape.commit()
doc.save("y.pdf")
```
![grafik](https://user-images.githubusercontent.com/8290722/228839790-dbbac5bc-2022-40a4-a376-4ba39cb6643f.png)

**Comment by sharon-br** [comment]
@JorjMcKie Thay you that indeed helped!
I went over the entire shape object documentation, however, I didn't notice that as relevant since I searched for references to "holes" or "interiors" and couldn't find any so I didn't notice it.
Anyways, thank you once again.
